sqlccl: merge IMPORT and RESTORE jobs

madelynnblue opened the pull request:
Previously the IMPORT statement would create both an IMPORT and a
RESTORE job (unless transform_only was specified). This created some
problems: it forced a temp directory to be specified and prevented
IMPORT from being a real job due to how the job code currently works.

Change IMPORT to instead directly ingest sstables as soon as they
are created, thus removing the need for the RESTORE job. This new
functionality is only available in distributed mode. Make distributed
the default since it is now well tested enough that we have high
enough confidence in it to flip that switch. Add a local option to
use local import instead. (If a user uses local, and thus must also
use transform, they will need to use a normal RESTORE job in order
to import the produced backups, which would require a CCL license.)

This change allows for some optimizations during sst generation,
like preallocating the table ID and generating data with it, thus
removing the need for any key rewriting, which has historically been
limiting speed factor during restores.

Remove transform_only and temp, but merge them into a new transform
option that performs only a transform.

This commit does not change IMPORT jobs to be handled by the registry,
which means that a failure could leave orphaned data.

While here, fix up subtests in TestImportStmt to use correct indexes
for jobs, and database and directory names.

Release note (enterprise): IMPORT CSV has had its required `temp`
parameter removed. In addition, the `transform_only` option has been
renamed to `transform` and now takes an argument specifying where to
store transformed CSV data (but keeps its behavior that, if specified,
only the transform part of the IMPORT will be performed). Finally,
IMPORT no longer creates a RESTORE job, but instead directly restores
data itself. For most uses of IMPORT, simply remove the `temp` option
from the query will achieve the same result as before.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/21490)
<!-- Reviewable:end -->

mjibson commented:
@tschottdorf Adding you to review my usage of AddSSTable to make sure it's correct since Dan is out.

mjibson commented:
Benchmark: 4 node geo-distributed cluster doing a 8-way split SF-10 TPCH IMPORT
old: 47 minutes

mjibson commented:
While running a sf-10 benchmark of this new code, I encountered this error after some 10s of minutes:

pq: unable to find store 1 in range r32:/{Table/51/1/50528/5-Max} [(n2,s2):1, (n4,s4):2, (n3,s3):3, next=4]

mjibson commented:
Review status: 0 of 3 files reviewed at latest revision, 1 unresolved discussion, all commit checks successful.

---

*[pkg/ccl/sqlccl/csv.go, line 1543 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/21490#-L36ZqbI0bmJlvxQtXeb:-L36ZqbI0bmJlvxQtXec:b7lvuoq) ([raw file](https://github.com/cockroachdb/cockroach/blob/a53d8d840e8380013719da81924342dc752e1a74/pkg/ccl/sqlccl/csv.go#L1543)):*
> ```Go
> 					log.Errorf(ctx, "failed to scatter span %s: %s", roachpb.PrettyPrintKey(nil, span.End), pErr)
> 				}
> 				if err := sp.db.AddSSTable(ctx, firstKey, lastKey, data); err != nil {
> ```

The restore code has a thing that computes stats after this. Do we need that here? I'm not familiar enough with this to know.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/21490)*
<!-- Sent from Reviewable.io -->

dt commented:
Reviewed 1 of 3 files at r1, 1 of 2 files at r2.
Review status: 2 of 3 files reviewed at latest revision, 2 unresolved discussions, all commit checks successful.

---

*[pkg/ccl/sqlccl/csv.go, line 1543 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/21490#-L36ZqbI0bmJlvxQtXeb:-L39hQ_Ae37Wv3uNKF9r:bxlbcsr) ([raw file](https://github.com/cockroachdb/cockroach/blob/a53d8d840e8380013719da81924342dc752e1a74/pkg/ccl/sqlccl/csv.go#L1543)):*
<details><summary><i>Previously, mjibson (Matt Jibson) wrote…</i></summary><blockquote>

The restore code has a thing that computes stats after this. Do we need that here? I'm not familiar enough with this to know.
</blockquote></details>

As far as correctness, I think `evalAddSStable` does the MVCC stat work that has to be done though, so this is more a nice-to-have, not a must-have. 

I can certainly see wanting to know how many rows I IMPORTed, and wouldn't mind the other numbers on index entries and byte size either. We could probably accumulate these during the aggregation / SST creation step above using a BulkOpSummary? Anyway, don't think it needs to block this change.

---

*[pkg/ccl/sqlccl/csv.go, line 1540 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/21490#-L39gPXZcRNXruOj7NIC:-L39gPXZcRNXruOj7NID:bqq5gaz) ([raw file](https://github.com/cockroachdb/cockroach/blob/3e9ff26a325ae166856dbaac70fc12907f5928fa/pkg/ccl/sqlccl/csv.go#L1540)):*
> ```Go
> 				}
> 				if _, pErr := client.SendWrapped(ctx, sp.db.GetSender(), scatterReq); pErr != nil {
> 					// TODO(dan): Unfortunately, Scatter is still too unreliable to
> ```

Comment is stale here (i.e. not in RESTORE). Could maybe just remove it.

We could mention something about in the future, trying to split and scatter on the coordinator so that the distsql collector can be scheduled on the actual ingesting node, instead of adding another network trip for the AddSSTable, but that's an optimization for later either way.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/21490)*
<!-- Sent from Reviewable.io -->

mjibson commented:
Review status: 2 of 3 files reviewed at latest revision, 2 unresolved discussions, all commit checks successful.

---

*[pkg/ccl/sqlccl/csv.go, line 1543 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/21490#-L36ZqbI0bmJlvxQtXeb:-L39qSyoU8ZFgW8Jc-tc:bb6c3tw) ([raw file](https://github.com/cockroachdb/cockroach/blob/a53d8d840e8380013719da81924342dc752e1a74/pkg/ccl/sqlccl/csv.go#L1543)):*
<details><summary><i>Previously, dt (David Taylor) wrote…</i></summary><blockquote>

As far as correctness, I think `evalAddSStable` does the MVCC stat work that has to be done though, so this is more a nice-to-have, not a must-have. 

I can certainly see wanting to know how many rows I IMPORTed, and wouldn't mind the other numbers on index entries and byte size either. We could probably accumulate these during the aggregation / SST creation step above using a BulkOpSummary? Anyway, don't think it needs to block this change.


</blockquote></details>

Yes, I'd like to do this in a followup commit.

---

*[pkg/ccl/sqlccl/csv.go, line 1540 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/21490#-L39gPXZcRNXruOj7NIC:-L39pftAACN-lvOopTPJ:bvqy5tq) ([raw file](https://github.com/cockroachdb/cockroach/blob/3e9ff26a325ae166856dbaac70fc12907f5928fa/pkg/ccl/sqlccl/csv.go#L1540)):*
<details><summary><i>Previously, dt (David Taylor) wrote…</i></summary><blockquote>

Comment is stale here (i.e. not in RESTORE). Could maybe just remove it.

We could mention something about in the future, trying to split and scatter on the coordinator so that the distsql collector can be scheduled on the actual ingesting node, instead of adding another network trip for the AddSSTable, but that's an optimization for later either way.
</blockquote></details>

I've changed RESTORE to IMPORT, but I've otherwise kept the comment the same because afaik it's still true, or at least untested. Yes, we could pre-split everything in the coordinator, but that would then take up 10s of minutes just splitting, because we'd have to do it all up front. Your proposal would also require the distsql plan to know which nodes are in the correct zone to be able to store the table, whereas the current code allows the scatter implementation to determine that.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/21490)*
<!-- Sent from Reviewable.io -->

dt commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: 2 of 3 files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/21490#-:-L3A0WFmbWI1ZZ6scobg:bnfp4nl)*
<!-- Sent from Reviewable.io -->

mjibson commented:
Running a SF-100 import had an error:

pq: ADDSSTable [/Table/51/1/10190243/2/0, /Table/51/1/10462464/1/0/NULL]: result is ambiguous (removing replica)

mjibson commented:
Current error:

pq: addsstable [/Table/51/1/4938598/4/0,/Table/51/1/6192549/6/0/NULL): command is too large: 75242286 bytes (max: 67108864)

mjibson commented:
Ok all bugs fixed I think. Changed some stuff around AddSSTables to retry if there was a split and to split up commands larger than max raft size. Tested with a 100GB SF IMPORT and it worked correctly over 8 hours.

tschottdorf commented:
Sorry to let this sit, Matt. I'll take a look first thing tomorrow.

On Sat, Jan 20, 2018, 7:18 PM Matt Jibson <notifications@github.com> wrote:

> Ok all bugs fixed I think. Changed some stuff around AddSSTables to retry
> if there was a split and to split up commands larger than max raft size.
> Tested with a 100GB SF IMPORT and it worked correctly over 8 hours.
>
> —
> You are receiving this because you were mentioned.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/cockroachdb/cockroach/pull/21490#issuecomment-359212967>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AE135JLdPbBMw2dDy-28xxW2-C9__PBaks5tMoJwgaJpZM4Rgb9R>
> .
>
-- 

-- Tobias

tschottdorf commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> on the AddSSTable part mod a comment.

---

Reviewed 1 of 2 files at r2, 4 of 4 files at r3.
Review status: all files reviewed at latest revision, 1 unresolved discussion, all commit checks successful.

---

*[pkg/ccl/sqlccl/csv.go, line 1559 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/21490#-L3TObh6Vh4MD5v1kY2H:-L3TObh6Vh4MD5v1kY2I:b2oa26a) ([raw file](https://github.com/cockroachdb/cockroach/blob/43b862171564079dc3a7f9561bd12c143f04ceaf/pkg/ccl/sqlccl/csv.go#L1559)):*
> ```Go
> 						log.Errorf(ctx, "failed to scatter span %s: %s", roachpb.PrettyPrintKey(nil, end), pErr)
> 					}
> 					if err := storageccl.AddSSTable(ctx, sp.db, firstKey, lastKey, data); err != nil {
> ```

`lastKey` is exclusive, right?  It can't be a key that's actually in the SSTable. Just double checking. If `lastKey` is already `actualLastKey.Next()` I think it's worth renaming the arguments to avoid confusion.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/21490#-:-L3TODRsCxCWZvj4nSnE:b-vgqfyx)*
<!-- Sent from Reviewable.io -->

madelynnblue committed:
sqlccl: merge IMPORT and RESTORE jobs

Previously the IMPORT statement would create both an IMPORT and a
RESTORE job (unless transform_only was specified). This created some
problems: it forced a temp directory to be specified and prevented
IMPORT from being a real job due to how the job code currently works.

Change IMPORT to instead directly ingest sstables as soon as they
are created, thus removing the need for the RESTORE job. This new
functionality is only available in distributed mode. Make distributed
the default since it is now well tested enough that we have high
enough confidence in it to flip that switch. Add a local option to
use local import instead. (If a user uses local, and thus must also
use transform, they will need to use a normal RESTORE job in order
to import the produced backups, which would require a CCL license.)

This change allows for some optimizations during sst generation,
like preallocating the table ID and generating data with it, thus
removing the need for any key rewriting, which has historically been
limiting speed factor during restores.

Remove transform_only and temp, but merge them into a new transform
option that performs only a transform.

This commit does not change IMPORT jobs to be handled by the registry,
which means that a failure could leave orphaned data.

While here, fix up subtests in TestImportStmt to use correct indexes
for jobs, and database and directory names.

While in the vicinity, fix a needless function call in a loop in
import.go.

Release note (enterprise): IMPORT CSV has had its required `temp`
parameter removed. In addition, the `transform_only` option has been
renamed to `transform` and now takes an argument specifying where to
store transformed CSV data (but keeps its behavior that, if specified,
only the transform part of the IMPORT will be performed). Finally,
IMPORT no longer creates a RESTORE job, but instead directly restores
data itself. For most uses of IMPORT, simply remove the `temp` option
from the query will achieve the same result as before.

mjibson commented:
Rebased for #21373 fix.

---

Review status: all files reviewed at latest revision, 1 unresolved discussion, all commit checks successful.

---

*[pkg/ccl/sqlccl/csv.go, line 1559 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/21490#-L3TObh6Vh4MD5v1kY2H:-L3UBwZTkkwz1fZOMrXq:b-6nqdy0) ([raw file](https://github.com/cockroachdb/cockroach/blob/43b862171564079dc3a7f9561bd12c143f04ceaf/pkg/ccl/sqlccl/csv.go#L1559)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

`lastKey` is exclusive, right?  It can't be a key that's actually in the SSTable. Just double checking. If `lastKey` is already `actualLastKey.Next()` I think it's worth renaming the arguments to avoid confusion.
</blockquote></details>

`lastKey` was exclusive, yes. However I've modified the code to make it more obvious what's going on. As a side effect we now only call .Next once per SSTable instead of on every KV, which was certainly a large amount of wasted CPU and memory copying.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/21490#-:-L3UBtR-P7UwsbijLG7v:b-7qm47k)*
<!-- Sent from Reviewable.io -->

mjibson commented:
@danhhz @tschottdorf Do I need to worry about MVCC stats after calling addsstable? I have no idea what those are and saw some code in restore dealing with them. Just want to make sure this is safe to merge if the tests pass.

danhhz commented:
AddSSTable handles updating them for you, so shouldn't be anything to worry about at the RESTORE/IMPORT level
